base/abstractarray.jl

```diff
@@ -343,6 +343,38 @@ function first(itr)
     x[1]
 end
 
+"""
+    first(predicate, coll)
+
+Get the first element of `coll` satisfying `predicate` wrapped in [`Some`](@ref).
+
+If no element of `coll` satisfies `predicate`, return `nothing`.
+
+!!! compat "Julia 1.6"
+    This method was added in Julia 1.6.
+
+# Examples
+```jldoctest
+julia> first(>(5), 1:10)
+Some(6)
+
+julia> isnothing(first(isodd, 2:2:10))
+true
+
+julia> something(first(iseven, [5, 3, 4, 2, 6, 8]))
+4
+
+julia> something(first(>(10), 1:10), 0)
+0
+```
+"""
+function first(predicate, itr)
+    for x in itr
+        predicate(x) && return Some(x)
```

Comment:
I'm not thrilled about `Some` here — I don't think we have a single API in base that returns `Some`.  That said, this is unlike the other find/search APIs in that they return indices or `RegexMatch`es or values from some other limited type set that does not include `Nothing`.

Reply:
I think I've convinced myself that at the very least this should be a separate function from `first`/`last`. Does `Iterators.first` (as suggested in OP) sound like a good name to you? In that case, we wouldn't have to create a predicate accepting version as it would compose with `Iterators.filter`.

In reality, I think all APIs that generically return an element of a collection but throw if the element isn't present should instead return `Union{Some{T},Nothing}`. This includes `first`, `last`, `getindex`, `reduce`, `foldl`, etc. This is obviously breaking; I should probably create a "Taking something seriously" issue to get feedback on that idea for julia 2.0.